src/main/java/org/opensearch/neuralsearch/query/NeuralSparseQuery.java

```diff
@@ -0,0 +1,92 @@
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Weight;
+
+/**
+ * Implementation of Query interface for type "hybrid". It allows execution of multiple sub-queries and collect individual
+ * scores for each sub-query.
+ */
+@AllArgsConstructor
+@Getter
+@NonNull
+public final class NeuralSparseQuery extends Query {
+
+    private Query currentQuery;
+    private final Query highScoreTokenQuery;
+    private final Query lowScoreTokenQuery;
+    private final Float rescoreWindowSizeExpansion;
+
+    @Override
+    public String toString(String field) {
+        return "NeuralSparseQuery("
+            + currentQuery.toString(field)
+            + ","
+            + highScoreTokenQuery.toString(field)
+            + ", "
+            + lowScoreTokenQuery.toString(field)
+            + ")";
+
+    }
+
+    public Query rewrite(IndexSearcher indexSearcher) throws IOException {
+        Query rewrittenCurrentQuery = currentQuery.rewrite(indexSearcher);
+        Query rewrittenFirstStepQuery = highScoreTokenQuery.rewrite(indexSearcher);
+        Query rewrittenSecondPhaseQuery = lowScoreTokenQuery.rewrite(indexSearcher);
+        if (rewrittenFirstStepQuery == highScoreTokenQuery && rewrittenSecondPhaseQuery == lowScoreTokenQuery) {
+            return this;
+        }
+        return new NeuralSparseQuery(rewrittenCurrentQuery, rewrittenFirstStepQuery, rewrittenSecondPhaseQuery, rescoreWindowSizeExpansion);
+    }
+
+    @Override
+    public void visit(QueryVisitor queryVisitor) {
+        QueryVisitor v = queryVisitor.getSubVisitor(BooleanClause.Occur.SHOULD, this);
+        currentQuery.visit(v);
+        highScoreTokenQuery.visit(v);
+        lowScoreTokenQuery.visit(v);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return sameClassAs(other) && equalsTo(getClass().cast(other));
+    }
+
+    private boolean equalsTo(NeuralSparseQuery other) {
+        return Objects.equals(currentQuery, other.currentQuery)
+            && Objects.equals(highScoreTokenQuery, other.highScoreTokenQuery)
+            && Objects.equals(lowScoreTokenQuery, other.lowScoreTokenQuery)
+            && Objects.equals(rescoreWindowSizeExpansion, other.rescoreWindowSizeExpansion);
+    }
+
+    @Override
+    public int hashCode() {
+        int h = classHash();
+        h = 31 * h + Objects.hashCode(highScoreTokenQuery) + Objects.hashCode(lowScoreTokenQuery) + Objects.hashCode(currentQuery) + Objects
+            .hashCode(rescoreWindowSizeExpansion);
+        return h;
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        return currentQuery.createWeight(searcher, scoreMode, boost);
+    }
+
+    public void extractLowScoreToken() {
+        this.currentQuery = highScoreTokenQuery;
+    }
```

Comment:
name of the function doesn't justify the operation performed by the function. Rename the function.

Reply:
Now change the name to setCurrentQueryToHighScoreTokenQuery().